### ISSUE TITLE
Current way masking training data creates issues with the Tokenizer

### ISSUE THREAD

**Issue by alealv** [Lightning-AI/litgpt#1567]
## Problem
The function [\_\_getitem__](https://github.com/Lightning-AI/litgpt/blob/main/litgpt/data/base.py#L77) in the `SFTDataset` computes the `input_ids` by doing `encode(prompt + response)`, but this could be different token than `encode(prompt), encode(response)` depending on the tokenizer.

## Example
I was fine-tuning [Mistral-7B-Instruct-v0.2](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2), which requires [CodeLlama](https://github.com/Lightning-AI/litgpt/blob/main/litgpt/prompts.py#L288) prompting style.

So, for example, if the response is a JSON, we would have:
```python
prompt = "[INST]...[INST]"
response = "{...}"
```

The problem arises when encoding `prompt + response` because, for the `Mistral-7B-Instruct-v0.2` tokenizer, we have:

```
      "]{": 4490,
      "]": 28793,
      "{": 28751,
```

So for training, we will get the token `4490`, but during generation, it will have `28793` as the last token. Therefore, it will fail to produce the expected output.

## How to fix 
We should set `input_ids = encode(prompt)+encode(response)`

### What operating system are you using?

Linux

### LitGPT Version

``` 
Version: 0.4.3
```

**Comment by rasbt** [comment]
Oh I see, thanks for raising that! So to summarize, there are cases where `encode(prompt)+encode(response) != `encode(prompt + response)`.

**Comment by alealv** [comment]
> Oh I see, thanks for raising that! So to summarize, there are cases where
> 
> ```
> encode(prompt) + encode(response) != encode(prompt + response).
> ```

exactly
